[receiver/dockerstats] Add optional resource attributes

marcsanmi opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
Add `container.image.id` and `container.command_line` as optional resource attributes, disabled by default.



**Link to tracking Issue:** <Issue number if applicable>

Tracking issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/21092

**Testing:** <Describe what testing was performed and which tests were added.>

Until this point, all resource attributes have been enabled by default, resulting in the coupling of the [test](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7bf5d66fae6c1c5360ec2875948f2c2adab6c44f/receiver/dockerstatsreceiver/receiver_test.go#L191) to the default configuration. In order to test the current implementation with the new resource attributes enabled & disabled, a minor refactoring was required in the `TestScrapeV2` test, allowing to pass different configurations for each test. 

Refer to the following commit 508dbee9aa9a22c1ca79f16bba5abeecceeb9c48 to check the refactor in the test.

**Documentation:** <Describe the documentation added.>

New resource attributes documentation was generated by `mdatagen`.

marcsanmi committed:
Add container imageID and command optional resource attributes

marcsanmi committed:
Refactor TestScrapeV2 to support different configs

jamesmoessis commented on receiver/dockerstatsreceiver/receiver_test.go:
```diff
@@ -169,32 +181,52 @@ func TestScrapeV2(t *testing.T) {
 	assert.NoError(t, err)
 	defer twoContainerEngineMock.Close()
 
+	singleContainerWithOptionalResourceAttributesEngineMock, err := dockerMockServer(&map[string]string{
```
You may have some refactoring do if https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21111 gets merged before this, since it changed the test case structure slightly. We can cross that bridge when we come to it though

marcsanmi replied:
Sure, when either this PR or https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21111  is merged, the other one will have to be rebased and apply any necessary changes.

jamesmoessis reviewed (approved):
LGTM

atoulme commented on receiver/dockerstatsreceiver/documentation.md:
```diff
@@ -678,8 +678,10 @@ Packets sent.
 
 | Name | Description | Values | Enabled |
 | ---- | ----------- | ------ | ------- |
+| container.command | The command executed by the container. | Any Str | false |
```
Please add this resource attribute to this table: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/container.md

marcsanmi replied:
I already created the PR https://github.com/open-telemetry/opentelemetry-specification/pull/3468.

atoulme commented on receiver/dockerstatsreceiver/receiver.go:
```diff
@@ -135,7 +135,10 @@ func (r *receiver) recordContainerStats(now pcommon.Timestamp, containerStats *d
-		metadata.WithContainerName(strings.TrimPrefix(container.Name, "/")))
+		metadata.WithContainerName(strings.TrimPrefix(container.Name, "/")),
+		metadata.WithContainerImageID(container.Image),
+		metadata.WithContainerCommand(strings.Join(container.Config.Cmd, ",")),
```
Take a look here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/process.md the `process.command_args` uses a []string representation. Would it make sense to follow that approach here too?

marcsanmi replied:
It would make sense since it's also what docker's API is returning. However, I was leveraging the `WithContainerCommand` func autogenerated with `mdatagen` and I think I cannot set the resource attribute in the `metadata.yaml` as a []string. Maybe I'm missing something and you can point me in the right direction, thanks!

marcsanmi replied:
I updated it from `container.command` to `container.command_line` to be reported as a string and to keep consistency with the conventions as defined here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/process.md#process (`process.command_line`). WDYT?

You can check the update in this commit f2b30c3e50f77cb1eb59b42426da5f4c0e928ae7.

marcsanmi committed:
Update container.command to container.command_line

atoulme commented:
Please rebase.

marcsanmi committed:
Fix rebase conflicts

marcsanmi commented:
Hi @jamesmoessis @atoulme, 
I rebased and fixed the conflicts regarding the tests introduced by https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21111. Could you review it again?

jamesmoessis commented on receiver/dockerstatsreceiver/receiver_test.go:
```diff
@@ -286,3 +320,39 @@ func dockerMockServer(urlToFile *map[string]string) (*httptest.Server, error) {
 		_, _ = rw.Write(data)
 	})), nil
 }
+
```
Thanks! this makes the tests more readable and workable with

jamesmoessis commented on receiver/dockerstatsreceiver/testdata/mock/two_containers/expected_metrics.yaml:
```diff
@@ -4,24 +4,24 @@ resourceMetrics:
         - key: container.runtime
           value:
             stringValue: docker
+        - key: container.hostname
```
Looks like these two metrics have switched order in the expected, which made the diff a bit harder to reason with. Do you know why this happened? I felt like this had deterministic order before. 

Do you think you can make it the same order? Because if we merge this it will create diffs in other PRs that have affected the same output

marcsanmi replied:
I just tried and regenerate all the mock files. We use this [line](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6c71da24ffbf823e3cf4530d91e32b596f7c6b77/receiver/dockerstatsreceiver/receiver_test.go#L250) to create them, without any manual intervention. So, I don't know why the order has changed...

I see there were some changes (https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22717, https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22916) recently in the golden package regarding writing metrics, maybe it's because of this.

marcsanmi committed:
Regenerate mock testdata

marcsanmi committed:
Regenerate data

marcsanmi committed:
regenerate data

marcsanmi commented:
Unrelated spark test failing. Apparently is going to be fixed in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23675.
